File: bindings/zig/src/status.zig
//! Status codes relayed from the C binding / libitb.

const std = @import("std");
const ffi = @import("ffi.zig").c;

/// Mirrors the C binding's `itb_status` numerically. Codes 11..13 are
/// the Triple blob-record / registry sentinels, 14..17 a reserved
/// block; 19..22 belong to the native Blob surface
/// (not wrapped here but relayed verbatim if libitb ever returns
/// them). The enum is non-exhaustive so an unknown future code is
/// carried through rather than trapping.
pub const Status = enum(c_uint) {
    ok = 0,
    bad_hash = 1,
    bad_key_bits = 2,
    bad_handle = 3,
    bad_input = 4,
    buffer_too_small = 5,
    encrypt_failed = 6,
    decrypt_failed = 7,
    seed_width_mix = 8,
    bad_mac = 9,
    mac_failure = 10,
    blob_malformed_recipe = 11,
    recipe_primitive_unknown = 12,
    unknown_profile = 13,
    reserved_14 = 14,
    reserved_15 = 15,
    reserved_16 = 16,
    reserved_17 = 17,
    blob_mode_mismatch = 19,
    blob_malformed = 20,
    blob_version_too_new = 21,
    blob_too_many_opts = 22,
    stream_truncated = 23,
    stream_after_final = 24,
    triple_closed = 25,
    profile_exists = 26,
    internal = 99,
    _,

    /// Wraps a raw C return code without validation.
    pub fn fromRaw(rc: c_uint) Status {
        return @enumFromInt(rc);
    }

    /// Short static label for the code (C-side string literal).
    pub fn label(self: Status) [:0]const u8 {
        return std.mem.span(ffi.itb_status_str(@intFromEnum(self)));
    }
};

File: bindings/nodejs/src/status.ts
// Status codes mirrored from the libitb3 C ABI
// (cmd/cshared/internal/capi/errors.go). Numeric values are stable
// across releases.

/** Integer status code returned by every libitb3 entry point. */
export enum Status {
  Ok = 0,
  BadHash = 1,
  BadKeyBits = 2,
  BadHandle = 3,
  BadInput = 4,
  BufferTooSmall = 5,
  EncryptFailed = 6,
  DecryptFailed = 7,
  SeedWidthMix = 8,
  BadMac = 9,
  MacFailure = 10,
  BlobMalformedRecipe = 11,
  RecipePrimitiveUnknown = 12,
  UnknownProfile = 13,
  BlobModeMismatch = 19,
  BlobMalformed = 20,
  BlobVersionTooNew = 21,
  BlobTooManyOpts = 22,
  StreamTruncated = 23,
  StreamAfterFinal = 24,
  TripleClosed = 25,
  ProfileExists = 26,
  Internal = 99,
}
